### ISSUE TITLE
open to patches that cleanup API parameter checks or documentation?

### ISSUE THREAD

**Issue by JWStaiert** [libsdl-org/SDL#7142]
How open are you to accepting patches that cleanup parameter checks (or lack of them) in the various user facing functions, or document the use cases for certain values? I’m teaching myself SDL by writing unit tests against the API and I’ve identified one problem so far having written tests only for the SDL_Surface functions.

Specifically, SDL_CreateSurfaceFrom in SDL3 will return without assigning minimalPitch to the pitch field when the caller passes pitch = 0. There doesn’t appear to be a use case where that behavior makes sense, especially given that minimum pitch is an implementation specific property AND  SDL_CalculatePitch is a static function in SDL_surface.c.

**Comment by slouken** [comment]
In general, we are totally open to cleanup and improvements. That particular API is not a bug. The function SDL_CreateSurfaceFrom() is designed for the application to create a surface from memory where the application knows the correct pitch. A pitch of zero actually means don't advance the surface pointer at all between rows, and allows a single row to be used to fill another surface. This is a very specialized usage, but SDL doesn't try to second guess what the application is trying to do with this function.

**Comment by JWStaiert** [comment]
Sam,

Thank you for responding, I’d let it go but I’ve confirmed zero pitch breaks at least one Blit function and I have yet to see documentation indicating surfaces initialized by SDL_CreateSurfaceFrom aren’t useable in blits.

Do you know firsthand if that was the original intent? From a c language perspective, a two dimensional array is laid out in memory the same way as a one dimensional array, so pitch is irrelevant.

Jason

**Comment by slouken** [comment]
I don't remember where it was necessary, but there was a reason why leaving 0 pitch was required. Again, since this surface is created from application supplied memory, the application should know what the pitch is and pass it in. SDL shouldn't make any assumptions about the values provided, and the applications shouldn't make any assumptions about what SDL will do with them.

TL;DR: pass the correct pitch for your memory :)

**Comment by JWStaiert** [comment]
According to the commit log that change was made to allow something called Maelstrom to use a surface as a staging area for textures. My gut say’s that was a very narrow use case that deserved its own API, especially give that SDL_CreateSurface does what you’d expect with zero pitch (sets it to minimum).

It seems like either that change needs refactoring, or the comments for SDL_CreateSurfaceFrom the function need updating to indicate that you end up with a broken surface when passing pitch zero.

Also, please suggest to slouken that closing issues with no more than a ‘don’t call that’ as an answer is bad etiquette. He/she created the change I mentioned above, so it looks like avoiding being responsible for their work.

Jason

**Comment by slouken** [comment]
Okay, clearly this was biting you somewhere, so I clarified the valid use case and return an error if you pass valid pixels and zero pitch, and added it to the automated tests so it's clear this is expected behavior.

Feel free to submit other bugs if you find inconsistencies that need fixing or clarification.

**Comment by JWStaiert** [comment]
Thank you, it says a lot that you changed your mind. I am trying to decide which project to support (SDL vs. SFML and others.) I was about to bail on SDL due to this issue - silently breaking an API between minor versions is not something I want to deal with.

Jason
